src/librustc_lint/types.rs

```diff
@@ -427,6 +454,14 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
                                 }
                             }
                         }
+
+                        if all_phantom {
+                            if check_kind.is_toplevel() {
+                                return FfiUnsafe(UNSIZED_STRUCT_MSG);
+                            }
```

Comment:
Sure, all this mess was mainly to distinguish the top-level phantom-data case from the zero-size struct. But @nagisa is right that not doing it simplifies a bunch this code, so we can also just use the same error message.